README.md
Rail Nation Bot
===

A bot which automatically plays the game [Rail Nation](https://www.rail-nation.com). For the moment it only collect all the bonuses from train station for you and your team, and also it's watching ads.

## Features

  - [x] collect all the bonuses for you & your team
  - [x] watch all the ads for you & your team
  - [x] store rewards (from watching ads) like free money, research points, free tickets
  - [x] stop the program with a key while running, or you can close it manually.
  - [ ] reload page when video it stucks
  - [ ] distribute research points
  - [ ] leveling a building in some priority order.

## Motivation

I started to make this bot, because I found it useful saving time. Let it collecting all the bonuses for you, meanwhile you relax, eat, make a shower, or pray.
In this way you advance in the game, and also have more time for yourself.

## Screenshots

The latest version uses the widget for overall improved speed, and improved performance when a bonus need to be collected, as it jumps to the top of the widget.

In **Screenshot 1** you can see an example of the widget.
![image](Screenshoot1.png "Screenshoot 1")

And in **Screenshot 2** you can see an example of video (ad), how it's looking.
![image](Screenshoot2.JPG "Screenshoot 2")

## Tech/framework used

**Built with**

   * [Python](https://docs.python.org/3/) ( version 3.8.3) running on Windows 7, 64 bit

**Modules used**
   * *os* .......................... -> changing the current folder for using images
   * *time* ..................... -> decreased CPU usage
   * *pyautogui* .......... -> click and collect bonuses. It has the main use on this program.
   * *cv2* and *numpy* -> for image recognition
   * *sys* ....................... -> for stopped the program
   * *pynput* ................... -> pause the program by key pressed
   * *datetime* ................ -> display the started and end time.

## Installation

   1. Put the folder in your name documents or wherever you want.  
Make sure you don't change the name folder `img`, or any image name or file path.

   2. Run [install.py](code/install.py) from the main **code** folder.

## Usage

1. Make sure you are in the main map window, in game, and if you are using chrome you are logged on in the browser. The program it was designed to work for `1920 x 1080` resolution, but it was added an option for calculating if there it's other resolution than default one.

2. Run the program.

3. When meeting a confirmation window of starting, make sure that the browser, is right underneath of that window. The program literally click the buttons.

### Stop the program

For stop it, all you need to do is pressing `esc` key from the keyboard, and within ~ 4 sec, it should display the next message, that the program was stopped by the user. If not, chances are that the script is in within 15 min waiting time, so you can stop the script easily pressing `ctrl + c` from the command window.

Alternatively, you can move the mouse to the top-left corner of your screen, until it stops, as this is the default behaviour of the *pyautogui* module.

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details about my code of conduit.

## Versioning
All changes can be see at [CHANGELOG.md](CHANGELOG.md)

&ndash; [V2.0](https://github.com/Cristishor201/RailNationBot/releases/tag/V2.0)

&ndash; [V1.1](https://github.com/Cristishor201/RailNationBot/releases/tag/V1.1)

&ndash; [V1.0](https://github.com/Cristishor201/RailNationBot/releases/tag/V1.0) - initial version

## Authors

  * Cristian Florescu  - [Cristishor201](https://github.com/Cristishor201) - initial work

## License

This project is licensed under the **MIT License** - see the [LICENSE.md](LICENSE) file for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [V2.0] - 2020-06-25
## Added
- Create *'ChangeLog.md'* for changes about this project
- configure if widget it's not install at all
- open the widget, if it's hidden
- update widget position, if it's misplaced by the default position
- Added update for all resolutions
- Added option if you setup wrong your resolution
- Printing the required modification, in case user want less time loading
- Added collected bonus before watching video from the widget
- Added a settings file, for all favorite modifications
- Added an emergency stop the script, by *'Esc'* button.
- Added a standby mode to wait 15 min, before trying again to collect
- Added a start up for multilingual

## Changed
- Changed watching video by how long their are, dynamically
- Changed all the functions, and moved them in an external file
- Changed to no longer unnecessary clicks, and more speed of the script
- Changed images quality and their path

## Updated
- Updated *'.gitignore'*
- A new relative and absolute position of the buttons

## Removed
- Removed technique of collecting bonuses in a row
- Removed technique of watched videos in a row
- Removed all the verifications if the video was finished or not
- Removed the logging / testing blocks of code
- Removed some unused images

## [1.1] - 2019-06-13
## Added
- update .gitignore
- replace the absolut path of the images's folder with the relative path
- infinite loop

## Fixed
- restart video if it stop
- in case bonus can not be collected, when the limit account was reached
- in case bonus already collected by someone else
- in case video is not loading
- resolved the error when white screen appear instead of video
- updated a broken link of an image
- resolved if 2 consecutive videos are not loading

## Removed
- turn off the sound, as you can turn off the sound of the tab

## [1.0] - 2019-05-18
### Added

- added the licence
- option to turn on / off the sound (on computer)
- collect bonuses in a row for you and your team
- collect bonuses in a row for you and your team
- collect bonuses and videos (only twice)
- don't open loterry ticket in case of winning it
- the documentation of using the application

[Unreleased]: https://github.com/Cristishor201/RailNationBot/compare/V1.1...HEAD
[1.1]: https://github.com/Cristishor201/RailNationBot/compare/V1.0...V1.1
[1.0]: https://github.com/Cristishor201/RailNationBot/releases/tag/V1.0


code/lang/lang.py
def lang(myLanguage):

    if myLanguage == 'en':
        assoc_1, assoc_2, continue_last = ['association_1.jpg','association_2.jpg','Continue_last.jpg']
        bonus_cant_already, lotery_free, lotery_bonus = ['bonus_cant-collected_already.jpg','lotery_free_2.jpg', 'lotery_free.jpg']
        bonus_you, video_present, secure_bonus, skip_buy_ticket = ['watch_bonus_you.jpg','video_present.jpg', 'secure_special_bonus.jpg', 'skip_buy_ticket.jpg']
        bonus_money, bonus_prestige, sudenly_error = ['image$.jpg', 'imagePP.jpg', 'sudenly_error.jpg']
    else:
        assoc_1, assoc_2, continue_last = [None]*3
        bonus_cant, bonus_already, lotery_free = [None]*3
        bonus_you, no_video, video_not_load, video_present, skip = [None]*5

    return [assoc_1, assoc_2, continue_last, bonus_cant_already, lotery_free, lotery_bonus, bonus_you, video_present, secure_bonus, skip_buy_ticket, bonus_money, bonus_prestige, sudenly_error]

if __name__ == "__main__":
    # usefull for debuging
    print(lang('en'))


code/railBonuses.py
import pyautogui, time, os
import cv2
import numpy as np
from datetime import datetime as date
from settings import setup
import resources.env as env
from lang.lang import lang
from pynput import keyboard

if __name__ == "__main__":
    ratio = [1, 1]
    ok = 0

    pyautogui.confirm(text="The program started at {}.".format(date.now().strftime("%H-%M")), title='Started', buttons=['OK'])
    print("Program started at", date.now().strftime("%H-%M"))

    Resolution, Language = setup()
    if Resolution is None:
        Resolution = [pyautogui.size().width, pyautogui.size().height]
        print("myResolution= {}".format(Resolution))

    env.init(Resolution, Language)
    env.update()
    print(date.now().strftime("%H-%M"))

    with keyboard.Listener(on_press=env._key_press) as listener:
        while True: # main loop
            env.refreshing()
            start = time.perf_counter()
            for i in range(3):
                # check for bonus first
                bonus = env.bonus()

                env.bonus_lotery()
                if env.f_stop():
                    break
                if bonus:
                    continue

                if env.video(i):
                    env.button(i)
                    while True:
                        if env.f_stop():
                            break
                        time.sleep(2)
                        if env.video_active() == False:
                            break
                    time.sleep(1.5)
                    env.errors()
                    env.watch_bonus_you()
                    while True:
                        if env.f_stop():
                            break
                        time.sleep(2)
                        if env.video_active() == False:
                            break
                    env.continue_you()
                    env.video_lotery()

            finish = time.perf_counter()
            if env.f_stop():
                break
            if round(finish-start, 2) < 2: # wait 15min
                print(date.now().strftime("%H-%M"),': waiting 15 min for bonus or video...')
                time.sleep(15*60)
                pyautogui.move(50, 0, duration=2) # escape from stand-by screen

        #while end
        listener.join()
    pyautogui.confirm(text="The program finished succesefully at {}.".format(date.now().strftime("%H-%M")), title='Finished', buttons=['OK'])
    print("Program stopped at", date.now().strftime("%H-%M"))


code/railBonus_1.0.py
import pyautogui, time, os
import cv2
import numpy as np

os.chdir('C://Users//Cristian//MyPythonScripts//railBonus//img') #change here with your path

stationBonus = [[1147, 722], [1503, 653], [1728, 489]] # [x,y]
stationVideo = [[1141, 663], [1492, 589], [1723, 433]]

watchYou = [1044, 788]
continueYou = [1006, 747]
theNext = [1009, 934]
theBack = [907, 933]
lotery = [1101, 311]

def soundOff():
    pyautogui.moveTo(1780, 1051, duration=1)
    pyautogui.click()
    pyautogui.moveTo(1790, 907, duration=0.7)
    pyautogui.click()

def imagesearch(image, precision=0.8):
    im = pyautogui.screenshot()
    #im.save('testarea.png') usefull for debugging purposes, this will save the captured region as "testarea.png"
    img_rgb = np.array(im)
    img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)
    template = cv2.imread(image, 0)
    template.shape[::-1]

    res = cv2.matchTemplate(img_gray, template, cv2.TM_CCOEFF_NORMED)
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(res)
    if max_val < precision:
        return [-1,-1]
    return max_loc

def appearIMG(file):
    pos = imagesearch(file)
    if pos[0] != -1:
        return True
    else:
        return False

def appearNext(file):
    pos = imagesearch(file)
    if pos[0] != -1:
        return True
    else:
        return False

def station(): # colect all rewards from one station
    time.sleep(2)
    for i in range(3):
        pyautogui.click(stationBonus[i][0], stationBonus[i][1], interval=1)
        
        if (appearIMG('lotery_free.jpg') == True): # free ticket from others
            pyautogui.click(lotery[0], lotery[1], interval=3)
            time.sleep(3)
        
        pyautogui.click(1160, 399, interval=0.3) # close reduction lotery tickets

def Bonus():
    time.sleep(2)
    for j in range(3):
        pyautogui.click(stationVideo[j][0], stationVideo[j][1], interval=1)
        time.sleep(35)

        # search for image
        if appearIMG('watch_bonus_you.jpg') == True:
            pyautogui.click(watchYou[0], watchYou[1], interval=1)
            time.sleep(35)
        if appearIMG('Continue_last.jpg') == True:
            pyautogui.click(continueYou[0], continueYou[1], interval=1)
            time.sleep(3)
        if (appearIMG('lotery_free.jpg') == True):
            pyautogui.click(lotery[0], lotery[1], interval=3)
            time.sleep(3)
        
        pyautogui.click(1160, 399, interval=0.3) # close reduction lotery tickets

soundOff() # close sound

#### Station 1 time (first)
station() # for me at the begining
pyautogui.moveTo(theNext[0], theNext[1], duration=1)
pyautogui.click()

while(appearIMG('close_mode_while.jpg') == True): # while I don't reach my station
    station()
    pyautogui.moveTo(theNext[0], theNext[1], duration=1)
    pyautogui.click()
    time.sleep(3)
    
    if(appearIMG('close_mode_while.jpg') == False):
        break

#### Watch video 1 time (first)
Bonus() # for me at the begining
pyautogui.moveTo(theNext[0], theNext[1], duration=1)
pyautogui.click()

while(appearIMG('close_mode_while.jpg') == True): # while I don't reach my station
    Bonus()
    pyautogui.moveTo(theNext[0], theNext[1], duration=1)
    pyautogui.click()
    time.sleep(3)
    
    if(appearIMG('close_mode_while.jpg') == False):
        break

#### Station 2 times (second)
station() # for me at the begining
pyautogui.moveTo(theNext[0], theNext[1], duration=1)
pyautogui.click()

while(appearNext('close_mode_while.jpg') == True): # while I don't reach my station
    station()
    pyautogui.moveTo(theNext[0], theNext[1], duration=1)
    pyautogui.click()
    time.sleep(3)
    
    if(appearIMG('close_mode_while.jpg') == False):
        break

soundOff() # turn on the sound
print("All bonuses has been collected successfully !")


code/resources/env.py
import pyautogui
import sys, os
import lang.lang as lang
import numpy as np
import cv2, time
from pynput import keyboard

screen = [1920, 1080]
ratio = [1, 1]
refresh = [375, 203]
buttons = [[170, 275], [270, 275], [370, 275]] # [x,y] default
regions = [[126, 261, 92, 25], [224, 261, 92, 25], [322, 261, 92, 25]] # [x, y, width, height]
center_region = [960, 540, 200, 200]

watchYou = [1050, 777]
continueYou = [1005, 740]
skipVideo = [1330, 710] # lucky guess
lotery = [1090, 330]

closeVideo = [1340, 380]
xButton = [1148, 412] # buy lottery ticket
closeButton = [960, 610] # limit exceed
restartVideo = [880, 707]
cancelVideo = [1033, 707]
errors = [885, 650] # sudenly error

_if_widget = [None, None, None] # present, open, position
name_images = [] # assoc_1, assoc_2, continue_last, bonus_cant_already, lotery_free, lotery_bonus, bonus_you, video_present, secure_bonus, skip_buy_ticket, bonus_money, bonus_prestige, sudenly_error
_stop = False

path = '' # for images

def init(resolution, Language='en'):
    res = [pyautogui.size().width, pyautogui.size().height]
    global ratio, screen, path, name_images
    if res != resolution:
        print("Be sincere and enter your real resolution in settings:", res)
        sys.exit() # stop the program
    ratio = [resolution[0]/screen[0], resolution[1]/screen[1]]
    screen = resolution
    path = "lang//" + Language + "//img"
    if ratio == [1,1]:
        print("Unchanged resolution")

    name_images = lang.lang(Language)
    os.chdir(os.path.abspath(path))

    if image_on_screen(name_images[0]): # assoc_1
        _if_widget[0] = True
        if image_on_screen(name_images[1]): # assoc_2
            _if_widget[1] = True
            x, y = image_on_screen(name_images[1], 0.8, True) # assoc_2 | position
            #print("[x,y]",[x,y])
            if [x,y] == [114, 195]:
                _if_widget[2] = True
                print("Widget is in the position")
            else:
                _if_widget[2] = False
        else:
            _if_widget[1] = False
            _if_widget[2] = False
    else:
        _if_widget[0] = False
        _if_widget[1] = False
        _if_widget[2] = False

def image_on_screen(img_str,precision=0.8, position=False ): # statica
    img = pyautogui.screenshot()

    img_rgb = np.array(img)
    img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)
    template = cv2.imread(img_str, 0)
    template.shape[::-1]

    res = cv2.matchTemplate(img_gray, template, cv2.TM_CCOEFF_NORMED)
    __, max_val, __, max_loc = cv2.minMaxLoc(res)
    if position == True: # if I want to take the position of image
        return max_loc
    if max_val < precision:
        return False # image is not founded
    else:
        return True # image match

def image_in_region(image, region, precision=0.8): # region screen option
    x, y, width, height = region
    img = pyautogui.screenshot(region=(x, y, width, height))

    img_rgb = np.array(img)
    img_gray = cv2.cvtColor(img_rgb, cv2.COLOR_BGR2GRAY)
    template = cv2.imread(image, 0)
    template.shape[::-1]

    res = cv2.matchTemplate(img_gray, template, cv2.TM_CCOEFF_NORMED)
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(res)
    if max_val < precision:
        return False # image is not founded
    else:
        return True  # image match

def update_ratio(val_list):
    global ratio
    return [int(val_list[0]*ratio[0]), int(val_list[1]*ratio[1])]

def update():
    global ratio, name_images
    py = [[74, 998],[74, 848],[447, 377],[92, 25]]
    if ratio != [1, 1]:
        ###### closeButton ??? - #exeed limit account & bonus collected already
        center_region[0:] = update_ratio(center_region[:2]) + update_ratio(center_region[2:])
        watchYou[0:] = update_ratio(watchYou)
        continueYou[0:] = update_ratio(continueYou)
        skipVideo[0:] = update_ratio(skipVideo)
        lotery[0:] = update_ratio(lotery)
        closeVideo[0:] = update_ratio(closeVideo)
        restartVideo[0:] = update_ratio(restartVideo)
        cancelVideo[0:] = update_ratio(cancelVideo)
        xButton[0:] = update_ratio(xButton)
        closeButton[0:] = update_ratio(closeButton)
        errors[0:] = update_ratio(errors)

        for i in range(len(py)):
            py[i][0] = int(py[i][0] * ratio[0])
            py[i][1] = int(py[i][1] * ratio[1])

    if _if_widget[0] == False: # present
        pyautogui.click(py[0]) # plus add widget button
        time.sleep(1)
        pyautogui.click(py[1]) # click association
        time.sleep(3)
        pyautogui.moveTo(py[2]) # corner right-down widget
        pyautogui.drag(0,10,1)
    elif _if_widget[1] == False: # open
        pyautogui.click(py[0])
        time.sleep(3)

    if _if_widget[2] == False or ratio != [1, 1]: # position or ratio
        x, y = image_on_screen(name_images[1], 0.8, True)
        refresh[0] = int((x + 261)* ratio[0]) # update refresh_x
        refresh[1] = int((y + 8)* ratio[1]) # update refresh_y
        x1, y1 = [int((x+56)*ratio[0]), int((y+80)*ratio[1])]
        x2, y2 = [int((x+12)*ratio[0]), int((y+66)*ratio[1])]
        for i in range(3):
            buttons[i] = [x1, y1] ; x1= int((x1 +100)*ratio[0])
            regions[i] = [x2, y2, py[3][0], py[3][1]] ; x2 = int((x2 + py[3][0] + 6)*ratio[0])

        print("refresh", refresh)
        print("regions:",regions)
        print("buttons", buttons)

def __region_capture(region, name='testarea'):
    img = pyautogui.screenshot(region=region)
    img.save(name + '.jpg')

def button(item):
    pyautogui.click(buttons[item])
    time.sleep(2)

def refreshing():
    pyautogui.click(refresh)
    time.sleep(2)

def watch_bonus_you():
    pyautogui.click(watchYou)
    time.sleep(2)

def region(item):
    return regions[item]

def video_active():
    if image_on_screen(name_images[8]):
        return True
    else:
        return False

def continue_you():
    if image_on_screen(name_images[2]) == True:
        pyautogui.click(continueYou)
        time.sleep(1)
    elif image_on_screen(name_images[9]) == True: # don't buy ticket
        pyautogui.click(xButton)
        time.sleep(1)

def video_lotery():
    if image_on_screen(name_images[4]) == True: # ticket from video watched
        pyautogui.click(xButton)
        time.sleep(1)

def bonus_lotery():
    if image_on_screen(name_images[5]) == True: # ticket won
        pyautogui.click(lotery)
        time.sleep(1)

def bonus_cant():
    if image_on_screen(name_images[3]) == True: #bonus can't OR bonus collected already
        pyautogui.click(closeButton)
        time.sleep(0.5)
        return True
    return False

def bonus():
    if image_in_region(name_images[10], regions[0]): # take bonus 1 (money)
        pyautogui.click(buttons[0])
        time.sleep(1)
        if bonus_cant():
            print("can't collect bonus")
            time.sleep(1) # for skiping standbye
        else:
            return True
    if image_in_region(name_images[10], regions[1]): # take bonus 2 (money)
        pyautogui.click(buttons[1])
        time.sleep(1)
        if bonus_cant():
            print("can't collect bonus")
            time.sleep(1) # for skiping standbye
        else:
            return True
    if image_in_region(name_images[11], regions[2]): # take bonus 3 (prestige)
        pyautogui.click(buttons[2])
        time.sleep(1)
        if bonus_cant():
            print("prestige collected already")
            time.sleep(1) # for skiping standbye
        else:
            return True
    return False

def video(item):
    if image_in_region(name_images[7], regions[item]):
        return True
    else:
        return False

def f_stop():
    return _stop

def _key_press(key):
    global _stop
    if key == keyboard.Key.esc:
        _stop = True
        return False

def errors():
    if image_on_screen(name_images[12]):
        pyautogui.click(errors)
        time.sleep(1)
